crates/federate/src/worker.rs

```diff
@@ -1,132 +1,192 @@
-static FOLLOW_ADDITIONS_RECHECK_DELAY: Lazy<chrono::TimeDelta> = Lazy::new(|| {
-  if *LEMMY_TEST_FAST_FEDERATION {
-    chrono::TimeDelta::try_seconds(1).expect("TimeDelta out of bounds")
-  } else {
-    chrono::TimeDelta::try_minutes(2).expect("TimeDelta out of bounds")
-  }
-});
-/// The same as FOLLOW_ADDITIONS_RECHECK_DELAY, but triggering when the last person on an instance
-/// unfollows a specific remote community. This is expected to happen pretty rarely and updating it
-/// in a timely manner is not too important.
-static FOLLOW_REMOVALS_RECHECK_DELAY: Lazy<chrono::TimeDelta> =
-  Lazy::new(|| chrono::TimeDelta::try_hours(1).expect("TimeDelta out of bounds"));
+/// SendManager --(has many)--> InstanceWorker --(has many)--> SendRetryTask
+///      |                            |                               |
+/// -----|------create worker -> loop activities--create task-> send activity
+///      |                            |                             vvvv
+///      |                            |                           fail or success
+///      |                            |           <-report result--   |
+///      |               <---order and aggrate results---             |
+///      |   <---send stats---        |                               |
+/// filter and print stats            |                               |
 pub(crate) struct InstanceWorker {
   instance: Instance,
-  // load site lazily because if an instance is first seen due to being on allowlist,
-  // the corresponding row in `site` may not exist yet since that is only added once
-  // `fetch_instance_actor_for_object` is called.
-  // (this should be unlikely to be relevant outside of the federation tests)
-  site_loaded: bool,
-  site: Option<Site>,
-  followed_communities: HashMap<CommunityId, HashSet<Url>>,
   stop: CancellationToken,
-  context: Data<LemmyContext>,
-  stats_sender: UnboundedSender<FederationQueueStateWithDomain>,
-  last_full_communities_fetch: DateTime<Utc>,
-  last_incremental_communities_fetch: DateTime<Utc>,
+  federation_lib_config: FederationConfig<LemmyContext>,
+  federation_worker_config: FederationWorkerConfig,
   state: FederationQueueState,
   last_state_insert: DateTime<Utc>,
+  pool: ActualDbPool,
+  inbox_collector: RealCommunityInboxCollector,
+  // regularily send stats back to the SendManager
+  stats_sender: UnboundedSender<FederationQueueStateWithDomain>,
+  // each HTTP send will report back to this channel concurrently
+  receive_send_result: mpsc::UnboundedReceiver<SendActivityResult>,
+  // this part of the channel is cloned and passed to the SendRetryTasks
+  report_send_result: mpsc::UnboundedSender<SendActivityResult>,
+  // activities that have been successfully sent but
+  // that are not the lowest number and thus can't be written to the database yet
+  successfuls: BinaryHeap<SendSuccessInfo>,
+  // number of activities that currently have a task spawned to send it
+  in_flight: i64,
```

Comment:
I dont see any reason for i64 here, even u8 should be enough.

Reply:
tbh i don't think in general it makes sense even thinking about these this. [unsigneds are always problematic](https://internals.rust-lang.org/t/revisiting-the-unsigned/1819), even for values that you expect to be positive. and an u8 probably doesn't even save any memory here on a 64bit platform. even if it did this is an int that's in memory max a few thousand times so the memory use difference is at max <10kB.

Reply:
The link only mentions that casting is easier, and that various languages use signed ints for array length. Neither of those is relevant here. But if you prefer then go with i8 instead. If we use i64 in the entire codebase because "its only a few kb difference" it will add up to a lot.

Reply:
i'm not going to put i8 because then someone is going to set their parallel sending to 1000 and then it's going to crash at random times

Reply:
and i'm not willing to write code either to cover that special case

Reply:
in production mode rust just wraps around, so for an unsigned int if you have 255 in there and add one it's going to be 0 which is going to cause impossible to diagnose problems. if you have a signed int and add 1 to 127 it's going to be -128 which is even weirder.

Reply:
i still disagree, but i've chenged it to i32 and added 4 ugly casts in random places around to make it still work.

i checked with size_of::<>() to check the size, and with an i32 the struct is now 832 bytes instead of 840. With an i8 it is still 832, so would be no difference.

Reply:
Of course you need to change the type in the config as well so it will reject too large values at startup.. Then theres only one extra cast needed, and in the same place theres already another cast.

Reply:
https://github.com/LemmyNet/lemmy/pull/4914